lib/std/crypto/25519/edwards25519.zig

```diff
@@ -0,0 +1,224 @@
+const std = @import("std");
+const fmt = std.fmt;
+
+/// Group operations over Edwards25519.
+pub const Edwards25519 = struct {
+    /// The underlying prime field.
+    pub const Fe = @import("field25519.zig").Fe;
+    /// Field arithmetic mod the order of the main subgroup.
+    pub const scalar = @import("scalar25519.zig");
+
+    x: Fe,
+    y: Fe,
+    z: Fe,
+    t: Fe,
+
+    is_base: bool = false,
+
+    /// Decode an Edwards25519 point from its compressed (Y+sign) coordinates.
+    pub fn fromBytes(s: [32]u8) !Edwards25519 {
+        const z = Fe.one();
+        const y = Fe.fromBytes(s);
+        var u = y.sq();
+        var v = u.mul(Fe.edwards25519d());
+        u = u.sub(z);
+        v = v.add(z);
+        const v3 = v.sq().mul(v);
+        var x = v3.sq().mul(v).mul(u).pow2523().mul(v3).mul(u);
+        const vxx = x.sq().mul(v);
+        const has_m_root = vxx.sub(u).isZero();
+        const has_p_root = vxx.add(u).isZero();
+        if ((@boolToInt(has_m_root) | @boolToInt(has_p_root)) == 0) {
+            return error.InvalidEncoding;
+        }
+        x.cMov(x.mul(Fe.sqrtm1()), 1 - @boolToInt(has_m_root));
+        x.cMov(x.neg(), @boolToInt(x.isNegative()) ^ (s[31] >> 7));
+        const t = x.mul(y);
+        return @as(Edwards25519, .{ .x = x, .y = y, .z = z, .t = t });
```

Comment:
Is this `@as` cast needed here? Wouldn't it be easier to just directly initialize a struct?

Reply:
Probably not needed! Let me change that!

Reply:
Done!